File: app/src/main/java/hegde/mahesh/avre/interpreter/BshInterpreter.java
package hegde.mahesh.avre.interpreter;

import java.io.OutputStream;
import java.io.PrintStream;

import bsh.EvalError;
import bsh.Interpreter;

public class BshInterpreter implements LanguageInterpreter {
    Interpreter ipr = new Interpreter();
    public BshInterpreter() {
        initializeInterpreter(ipr);
    }

    @Override
    public void setVariable(String variable, Object value) throws SnippetEvalException {
        try {
            ipr.set(variable, value);
        } catch (EvalError e) {
            throw convertEvalError(e);
        }
    }

    @Override
    public void setOutputStream(OutputStream out) {
        ipr.setOut(new PrintStream(out));
    }

    @Override
    public void setErrorStream(OutputStream err) {
        ipr.setErr(new PrintStream(err));
    }

    @Override
    public Object eval(String snippet) throws SnippetEvalException {
        try {
            return ipr.eval(snippet);
        } catch (EvalError e) {
            throw convertEvalError(e);
        }
    }

    @Override
    public String[] getCompletions(String partialInput, int position) {
        return new String[]{};
    }

    @Override
    public void reset() {
        ipr = new Interpreter();
        initializeInterpreter(ipr);
    }

    private void initializeInterpreter(Interpreter ignored) {
        // Eval startup file
    }

    private static SnippetEvalException convertEvalError(EvalError e) {
        int lineNum = e.getErrorLineNumber();
        int colNum = -1;
        Throwable cause = e.getCause();
        if (cause == null) {
            cause = e;
        }
        return new SnippetEvalException(lineNum, colNum, cause);
    }
}

File: app/src/main/java/hegde/mahesh/avre/interpreter/BshInterpreter.kt
package hegde.mahesh.avre.interpreter

import bsh.EvalError
import bsh.Interpreter
import java.io.OutputStream
import java.io.PrintStream

class BshInterpreter : LanguageInterpreter {
    private var ipr = Interpreter()

    @Throws(SnippetEvalException::class)
    override fun setVariable(variable: String, value: Any?) {
        try {
            ipr[variable] = value
        } catch (e: EvalError) {
            throw convertEvalError(e)
        }
    }

    override fun setOutputStream(out: OutputStream) {
        ipr.out = PrintStream(out)
    }

    override fun setErrorStream(err: OutputStream) {
        ipr.err = PrintStream(err)
    }

    @Throws(SnippetEvalException::class)
    override fun eval(snippet: String): Any {
        return try {
            ipr.eval(snippet)
        } catch (e: EvalError) {
            throw convertEvalError(e)
        }
    }

    override fun getCompletions(partialInput: String, position: Int): Array<String> {
        return arrayOf()
    }

    override fun reset() {
        ipr = Interpreter()
    }

    companion object {
        private fun convertEvalError(e: EvalError): SnippetEvalException {
            val lineNum = e.errorLineNumber
            var cause = e.cause
            if (cause == null) {
                cause = e
            }
            return SnippetEvalException(lineNum, cause)
        }
    }
}